src/shims/env.rs

```diff
@@ -20,15 +21,32 @@ impl EnvVars {
         ecx: &mut InterpCx<'mir, 'tcx, Evaluator<'tcx>>,
         excluded_env_vars: Vec<String>,
     ) {
+        let mut vars = Vec::new();
         if ecx.machine.communicate {
+                // Put each environment variable pointer in `EnvVars`, collect pointers.
             for (name, value) in env::vars() {
                 if !excluded_env_vars.contains(&name) {
                     let var_ptr =
                         alloc_env_var_as_c_str(name.as_ref(), value.as_ref(), ecx);
                     ecx.machine.env_vars.map.insert(OsString::from(name), var_ptr);
+                    vars.push(var_ptr);
                 }
             }
         }
+        // Make an array with all these pointers, in the Miri memory.
+        let tcx = ecx.tcx;
+        let environ_layout =
+            ecx.layout_of(tcx.mk_array(tcx.mk_imm_ptr(tcx.types.u8), vars.len() as u64)).unwrap();
+        let environ_place = ecx.allocate(environ_layout, MiriMemoryKind::Env.into());
+        for (idx, var) in vars.into_iter().enumerate() {
+            let place = ecx.mplace_field(environ_place, idx as u64).unwrap();
+            ecx.write_scalar(var, place.into()).unwrap();
+        }
+        ecx.memory.mark_immutable(environ_place.ptr.assert_ptr().alloc_id).unwrap();
+        // A pointer to that place corresponds to the `environ` static.
+        let environ_ptr = ecx.force_ptr(environ_place.ptr).unwrap();
+        let environ_alloc = ecx.memory.get_raw(environ_ptr.alloc_id).unwrap().clone();
```

Comment:
This is suboptimal as the allocation will be left in `Memory` forever even if it's never going to get used. I'm not sure if this is feasible, but you may be able to create the alloc with `Allocation::undef(size, align)` and then write to it directly instead of going through all the extra accessors above.

Reply:
Having the allocation last forever makes sense to me, it's a static after all.

We'll also need to update this allocation when the environment changes (whenever `environ_place` above gets reallocated).